Optimize looking up keys by raw reference in hash containers whose key is a non-nullable smart pointer type

cdumez opened the pull request:
#### 38bcd3245042be3cb678e0242a6ab95d4b1a9688
<pre>
Optimize looking up keys by raw reference in hash containers whose key is a non-nullable smart pointer type
<a href="https://bugs.webkit.org/show_bug.cgi?id=266166">https://bugs.webkit.org/show_bug.cgi?id=266166</a>

Reviewed by Darin Adler.

Optimize looking up keys by raw reference in hash containers whose key is a
non-nullable smart pointer type (such as Ref&lt;&gt; / WeakRef&lt;&gt; / CheckedRef&lt;&gt;).

I noticed while adopting WeakRef&lt;&gt; more in our containers that hash container
functions can be passed a raw C++ reference. This is convenient but actually
generates unnecessary refcounting churn for many of these functions because
it creates a WeakRef. Creating a WeakRef is unnecessary for lookup functions
such as contains() / find() and removal functions such as remove() / take().

We already had overloads in place for these functions when passing in
a raw C++ pointer and the key is a smart pointer type to avoid refcounting
churn. However, we didn&apos;t have such optimization when passing a raw C++
reference, which is useful when the key type is a non-nullable smart pointer.

This patch adds to our hash containers function overloads that take in a
raw C++ reference, similar to the ones that take in a raw C++ pointer. These
overloads are only available when the key is a non-nullable smart pointer
type.

I also make an improvement to our existing overloads taking in a raw C++
pointer so that they no longer require a non-const pointer. There is no
reason to require a non-const pointer just for pointer hashing and pointer
comparison. I also noticed this issue when adopting more smart pointers in
containers.

* Source/JavaScriptCore/runtime/Identifier.h:
(JSC::IdentifierRepHash::hash):
* Source/WTF/wtf/CheckedPtr.h:
* Source/WTF/wtf/CheckedRef.h:
* Source/WTF/wtf/CompactPtr.h:
* Source/WTF/wtf/ConcurrentPtrHashSet.h:
* Source/WTF/wtf/GetPtr.h:
* Source/WTF/wtf/HashCountedSet.h:
(WTF::Traits&gt;::find):
(WTF::Traits&gt;::find const):
(WTF::Traits&gt;::contains const):
(WTF::Traits&gt;::count const):
(WTF::Traits&gt;::remove):
* Source/WTF/wtf/HashFunctions.h:
* Source/WTF/wtf/HashMap.h:
(WTF::Y&gt;::find):
(WTF::Y&gt;::find const):
(WTF::Y&gt;::contains const):
(WTF::Y&gt;::inlineGet const):
(WTF::Y&gt;::get const):
(WTF::Y&gt;::remove):
(WTF::Y&gt;::take):
* Source/WTF/wtf/HashSet.h:
(WTF::TableTraits&gt;::find const):
(WTF::TableTraits&gt;::contains const):
(WTF::TableTraits&gt;::remove):
(WTF::TableTraits&gt;::take):
* Source/WTF/wtf/ListHashSet.h:
(WTF::U&gt;::find):
(WTF::U&gt;::find const):
(WTF::U&gt;::contains const):
(WTF::U&gt;::insertBefore):
(WTF::U&gt;::remove):
* Source/WTF/wtf/Packed.h:
* Source/WTF/wtf/Ref.h:
* Source/WTF/wtf/RefPtr.h:
* Source/WTF/wtf/RetainPtr.h:
* Source/WTF/wtf/SchedulePair.h:
(WTF::SchedulePairHash::hash):
(WTF::SchedulePairHash::equal):
* Source/WTF/wtf/SignedPtr.h:
* Source/WTF/wtf/UniqueRef.h:
* Source/WTF/wtf/WeakPtr.h:
(WTF::WeakPtr::WeakPtr):
* Source/WTF/wtf/WeakRef.h:
(WTF::WeakRef::releaseImpl):
(WTF::WeakRefHashTraits::take):
* Source/WTF/wtf/glib/GRefPtr.h:
* Source/WTF/wtf/text/StringHash.h:
(WTF::StringHash::hash):
(WTF::ASCIICaseInsensitiveHash::hash):
* Source/WebCore/dom/GCReachableRef.h:
* Source/WebCore/page/SecurityOriginHash.h:
(WebCore::SecurityOriginHash::hash):
(WebCore::SecurityOriginHash::equal):
* Source/WebCore/platform/win/COMPtr.h:
* Source/WebCore/rendering/FloatingObjects.h:
(WebCore::FloatingObjectHashFunctions::hash):
(WebCore::FloatingObjectHashFunctions::equal):
* Source/WebCore/rendering/GlyphDisplayListCache.h:
(WebCore::GlyphDisplayListCacheEntryHash::hash):
(WebCore::GlyphDisplayListCacheEntryHash::equal):
* Source/WebKit/UIProcess/API/cpp/WKRetainPtr.h:
* Tools/TestWebKitAPI/Tests/WTF/WeakPtr.cpp:
(TestWebKitAPI::TestType::resetDidUpdateImplRefCount):
(TestWebKitAPI::TestType::didUpdateImplRefCount const):
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/271853@main">https://commits.webkit.org/271853@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f8fafd691f67a6a3bc7f8385251370dc3b4f2fa8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29849 "22 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8512 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/31163 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32360 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26990 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/30504 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10694 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5776 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/27045 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/30140 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/7131 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/25465 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/6083 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/6252 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/26552 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33695 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/25637 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/27227 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/26982 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/32426 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/30032 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/6168 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4362 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/30210 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7912 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/36418 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/6918 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7856 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3851 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6697 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/CheckedPtr.h:
```diff
@@ -264,6 +264,7 @@ class CheckedPtr {
 template <typename T, typename PtrTraits>
 struct GetPtrHelper<CheckedPtr<T, PtrTraits>> {
     typedef T* PtrType;
+    typedef T UnderlyingType;
```
In new code we should use `using` rather than `typedef`. Noticed this many places in this patch.

cdumez replied:
Fixed.

darinadler commented on Source/WebCore/rendering/GlyphDisplayListCache.h:
```diff
@@ -91,8 +91,11 @@ inline void add(Hasher& hasher, const GlyphDisplayListCacheEntry& entry)
+    static unsigned hash(const GlyphDisplayListCacheEntry* entry) { return computeHash(*entry); }
     static unsigned hash(const CheckedPtr<GlyphDisplayListCacheEntry>& entry) { return computeHash(*entry); }
     static bool equal(const CheckedPtr<GlyphDisplayListCacheEntry>& a, const CheckedPtr<GlyphDisplayListCacheEntry>& b) { return a.get() == b.get(); }
+    static bool equal(const CheckedPtr<GlyphDisplayListCacheEntry>& a, const GlyphDisplayListCacheEntry* b) { return a.get() == b; }
```
I’d exxpect we would be able to just write `a == b` because of overloads in `CheckePtr`’s `==` implementation.

cdumez replied:
Fixed.

cdumez committed:
Optimize looking up keys by raw reference in hash containers whose key is a non-nullable smart pointer type
https://bugs.webkit.org/show_bug.cgi?id=266166

Reviewed by Darin Adler.

Optimize looking up keys by raw reference in hash containers whose key is a
non-nullable smart pointer type (such as Ref<> / WeakRef<> / CheckedRef<>).

I noticed while adopting WeakRef<> more in our containers that hash container
functions can be passed a raw C++ reference. This is convenient but actually
generates unnecessary refcounting churn for many of these functions because
it creates a WeakRef. Creating a WeakRef is unnecessary for lookup functions
such as contains() / find() and removal functions such as remove() / take().

We already had overloads in place for these functions when passing in
a raw C++ pointer and the key is a smart pointer type to avoid refcounting
churn. However, we didn't have such optimization when passing a raw C++
reference, which is useful when the key type is a non-nullable smart pointer.

This patch adds to our hash containers function overloads that take in a
raw C++ reference, similar to the ones that take in a raw C++ pointer. These
overloads are only available when the key is a non-nullable smart pointer
type.

I also make an improvement to our existing overloads taking in a raw C++
pointer so that they no longer require a non-const pointer. There is no
reason to require a non-const pointer just for pointer hashing and pointer
comparison. I also noticed this issue when adopting more smart pointers in
containers.

* Source/JavaScriptCore/runtime/Identifier.h:
(JSC::IdentifierRepHash::hash):
* Source/WTF/wtf/CheckedPtr.h:
* Source/WTF/wtf/CheckedRef.h:
* Source/WTF/wtf/CompactPtr.h:
* Source/WTF/wtf/ConcurrentPtrHashSet.h:
* Source/WTF/wtf/GetPtr.h:
* Source/WTF/wtf/HashCountedSet.h:
(WTF::Traits>::find):
(WTF::Traits>::find const):
(WTF::Traits>::contains const):
(WTF::Traits>::count const):
(WTF::Traits>::remove):
* Source/WTF/wtf/HashFunctions.h:
* Source/WTF/wtf/HashMap.h:
(WTF::Y>::find):
(WTF::Y>::find const):
(WTF::Y>::contains const):
(WTF::Y>::inlineGet const):
(WTF::Y>::get const):
(WTF::Y>::remove):
(WTF::Y>::take):
* Source/WTF/wtf/HashSet.h:
(WTF::TableTraits>::find const):
(WTF::TableTraits>::contains const):
(WTF::TableTraits>::remove):
(WTF::TableTraits>::take):
* Source/WTF/wtf/ListHashSet.h:
(WTF::U>::find):
(WTF::U>::find const):
(WTF::U>::contains const):
(WTF::U>::insertBefore):
(WTF::U>::remove):
* Source/WTF/wtf/Packed.h:
* Source/WTF/wtf/Ref.h:
* Source/WTF/wtf/RefPtr.h:
* Source/WTF/wtf/RetainPtr.h:
* Source/WTF/wtf/SchedulePair.h:
(WTF::SchedulePairHash::hash):
(WTF::SchedulePairHash::equal):
* Source/WTF/wtf/SignedPtr.h:
* Source/WTF/wtf/UniqueRef.h:
* Source/WTF/wtf/WeakPtr.h:
(WTF::WeakPtr::WeakPtr):
* Source/WTF/wtf/WeakRef.h:
(WTF::WeakRef::releaseImpl):
(WTF::WeakRefHashTraits::take):
* Source/WTF/wtf/glib/GRefPtr.h:
* Source/WTF/wtf/text/StringHash.h:
(WTF::StringHash::hash):
(WTF::ASCIICaseInsensitiveHash::hash):
* Source/WebCore/dom/GCReachableRef.h:
* Source/WebCore/page/SecurityOriginHash.h:
(WebCore::SecurityOriginHash::hash):
(WebCore::SecurityOriginHash::equal):
* Source/WebCore/platform/win/COMPtr.h:
* Source/WebCore/rendering/FloatingObjects.h:
(WebCore::FloatingObjectHashFunctions::hash):
(WebCore::FloatingObjectHashFunctions::equal):
* Source/WebCore/rendering/GlyphDisplayListCache.h:
(WebCore::GlyphDisplayListCacheEntryHash::hash):
(WebCore::GlyphDisplayListCacheEntryHash::equal):
* Source/WebKit/UIProcess/API/cpp/WKRetainPtr.h:
* Tools/TestWebKitAPI/Tests/WTF/WeakPtr.cpp:
(TestWebKitAPI::TestType::resetDidUpdateImplRefCount):
(TestWebKitAPI::TestType::didUpdateImplRefCount const):
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/271853@main
